cache/jemalloc_nodump_allocator.cc

```diff
@@ -0,0 +1,135 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "cache/jemalloc_nodump_allocator.h"
+
+#include <string>
+
+#include "util/string_util.h"
+
+namespace rocksdb {
+
+#ifdef ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
+
+std::atomic<extent_alloc_t*> JemallocNodumpAllocator::original_alloc_{nullptr};
+
+JemallocNodumpAllocator::JemallocNodumpAllocator(
+    unsigned arena_index, int flags, std::unique_ptr<extent_hooks_t>&& hooks)
+    : arena_index_(arena_index), flags_(flags), hooks_(std::move(hooks)) {
+  assert(arena_index != 0);
+}
+
+void* JemallocNodumpAllocator::Allocate(size_t size) {
+  return mallocx(size, flags_);
+}
+
+void JemallocNodumpAllocator::Deallocate(void* p) { dallocx(p, flags_); }
+
+void* JemallocNodumpAllocator::Alloc(extent_hooks_t* extent, void* new_addr,
+                                     size_t size, size_t alignment, bool* zero,
+                                     bool* commit, unsigned arena_ind) {
+  extent_alloc_t* original_alloc =
+      original_alloc_.load(std::memory_order_relaxed);
+  assert(original_alloc != nullptr);
+  void* result = original_alloc(extent, new_addr, size, alignment, zero, commit,
+                                arena_ind);
+  if (result != nullptr) {
+    int ret = madvise(result, size, MADV_DONTDUMP);
+    if (ret != 0) {
+      fprintf(
+          stderr,
+          "JemallocNodumpAllocator failed to set MADV_DONTDUMP, error code: %d",
+          ret);
+      assert(false);
+    }
+  }
+  return result;
+}
+
+JemallocNodumpAllocator::~JemallocNodumpAllocator() {
+  assert(arena_index_ != 0);
+  std::string key = "arena." + ToString(arena_index_) + ".destroy";
```

Comment:
Do we actually need this functionality? Destroying an arena will also release all the memory managed by the arena, which should be done carefully. If the allocator is only created once and never released, it's fine not destroying it at shutdown.

Reply:
On one hand, it is pretty safe to release all memory at this point. When the allocator destruct, it means the `Cache` object referencing the allocator is destroyed, which means the cache is not used any more.

On the other hand, the allocator is not necessary only destroy on shutdown. Our API allow user to open a RocksDB instance then close it and then continue with other stuff. Leaking the memory usage would be a bad idea.

Reply:
I'll add a comment state the allocator should not be destruct while there's allocated memory in use.